x/superfluid/keeper/intermediary_account.go

```diff
@@ -107,10 +110,76 @@ func (k Keeper) SetIntermediaryAccount(ctx sdk.Context, acc types.SuperfluidInte
 	prefixStore.Set(acc.GetAccAddress(), bz)
 }
 
-func (k Keeper) DeleteIntermediaryAccount(ctx sdk.Context, address sdk.AccAddress) {
+func (k Keeper) DeleteAllEmptyIntermediaryAccounts(ctx sdk.Context) {
+	itermedairyAccounts := k.GetAllIntermediaryAccounts(ctx)
+	for _, intermediaryAccount := range itermedairyAccounts {
+		k.DeleteIntermediaryAccountIfNoDelegation(ctx, intermediaryAccount)
+	}
+}
+
+// DeleteIntermediaryAccount deletes given intermediary account from store
+// Note that intermediary account is highly related to staking and delgation, and this
+// method should be used with caution.
+func (k Keeper) DeleteIntermediaryAccountIfNoDelegation(ctx sdk.Context, intermedairyAcc types.SuperfluidIntermediaryAccount) {
+	// check if any delegations or intermediary connections exist
+	if k.IntermediaryAccountDelegationsExists(ctx, intermedairyAcc) {
+		return
+	}
+
 	store := ctx.KVStore(k.storeKey)
+
+	// store for intermediary account
 	prefixStore := prefix.NewStore(store, types.KeyPrefixIntermediaryAccount)
-	prefixStore.Delete(address)
+	prefixStore.Delete(intermedairyAcc.GetAccAddress())
+}
+
+// IntermediaryAccountDelegationsExists returns true if the gien intermediary account has any delegations.
+// We check this by
+// - using staking keeper to check if actual delegations exist,
+// - checking if there are any intermediary account connection remaining to the intermediaryAcc
+// - if there's no synthetic lock with the intermediaryAccount.Denom + intermediaryAccount.ValAddr combination
+func (k Keeper) IntermediaryAccountDelegationsExists(ctx sdk.Context, intermedairyAcc types.SuperfluidIntermediaryAccount) (delegations bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	// we first check if the intermediary account does not have any connections
+	// store for intermediary account connection
+	intermediaryAccConnectionPrefixStore := prefix.NewStore(store, types.KeyPrefixLockIntermediaryAccAddr)
+	iterator := intermediaryAccConnectionPrefixStore.Iterator(nil, nil)
+
+	intermediaryConnectionExists := false
+
+	for ; iterator.Valid(); iterator.Next() {
+		if sdk.AccAddress(iterator.Value()).Equals(intermedairyAcc.GetAccAddress()) {
+			intermediaryConnectionExists = true
+		}
+	}
+
+	if intermediaryConnectionExists {
+		return true
+	}
+
+	// now check and verify that we don't have any delegations
+	_, found := k.sk.GetDelegation(ctx, intermedairyAcc.GetAccAddress(), sdk.ValAddress(intermedairyAcc.ValAddr))
+	if found {
+		return found
+	}
+
+	// check that the there's no synth denom with the synthdenom from the intermediary account
+	locks := k.lk.GetLocksLongerThanDurationDenom(ctx, intermedairyAcc.Denom, time.Second)
+	for _, lock := range locks {
+		// check if there are bonded synth locks with the synthdenom
+		synthLock, _ := k.lk.GetSyntheticLockup(ctx, lock.ID, stakingSyntheticDenom(intermedairyAcc.Denom, intermedairyAcc.ValAddr))
```

Comment:
I think err check might be missed here

Reply:
This is intentional, `GetSyntheticLockup` returns an error when it doesnt have the given lock, but in our case, this is not an error

We can also add a comment for this if this happens to be confusing.

I agree with you that this is bad API usage, an idea I would like to move forward to  after getting this PR merged is adding a method `HasSyntheticLockup` in the lockup keeper that returns a boolean for usage like this(we have other instances that uses `GetSyntheticLockup` like this whilst ignoring error, just to check if we have a synthetic lock)